### PR TITLE
ATR-572: Add Acuminator 2.3 RN and new diagnostic descriptions

### PR DESCRIPTION
**Pull request by EAndrosova**
@dropsonic Please review, but do not merge until norification. I am waiting the editor review as well.

### CODE REVIEW

File: docs/ReleaseNotes.md [hunk 1]
```diff
@@ -1,6 +1,28 @@
 # Acuminator Release Notes
 This document provides information about fixes, enhancements, and key features that are available in Acuminator.
 
+## Acuminator 2.3
+Acuminator 2.3 includes the bug fixes and enhancements described in this section, as well as the features that have been implemented in previous versions.
+
+### Enhancements
+In Acuminator 2.3, the following enhancements have been implemented:
+ - The [PX1036](diagnostics/PX1036.md) diagnostic has been added to check the correct naming of key field declarations in PK or FK API.
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:9]: @EAndrosova the check implemented only for primary keys

File: docs/ReleaseNotes.md [hunk 2]
```diff
@@ -1,6 +1,28 @@
...
+ - The [PX1035](diagnostics/PX1035.md) diagnostic has been added to check and warn of multiple primary key declarations.
+ - The [PX1033] (diagnostics/PX1033.md) and [PX1034] (diagnostics/PX1034.md) diagnostics has been added to suggest adding PK or FK definition to a DAC.
+ - The [PX1016](diagnostics/PX1016.md) diagnostic has been added to check whether a graph extension has the IsActive method implemented.
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:12]: it checks both graph extensions and DAC extensions

File: docs/ReleaseNotes.md [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+ - The Code Map has been improved as follows:
+   - The Code Map now indicates whether a graph is for mass processing operations and whether a graph node corresponds to a graph or a graph extension.
+   - The Code Map now indicates whether a DAC field is auto-incremented (has the PXDBIdentityt attribute).
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:15]: It's not PXDBIdentity attribute, it is AutoNumberAttribute

File: docs/ReleaseNotes.md [hunk 4]
```diff
@@ -1,6 +1,28 @@
...
+   - The following complementary information is now displayed in the Code Map about a graph view member: Whether the view is for mass processing operations, for setup form, or for filtering, whether the view is derived fom the base PXSelect or PXSelectReadOnly types.
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:16]: Imho it's better to say graph view instead of graph view member. I would also change the enumeration of indicators into something like a list:
Code Map now displays extra information about the graph views. You can see indicators for:
- Views for mass processing operations. // We frequently use term processing views
- Setup view //Views used to query for setup settings, I'm not sure if we need to mention setup forms here because these views are used by many data entry screens
- Filter views  //It's another established term, views derived from PXFilter
- Custom views derived from base PXSelect-types  // This are special custom views which were used some time ago to share common logic for different screens, for example, approval functionality. 
- Views derived from PXSelectReadOnly. // These views are not merged with graph cache, therefore they are marked as **Uncached** in the Code Map. This could be mentioned in the notes

File: docs/ReleaseNotes.md [hunk 5]
```diff
@@ -1,6 +1,28 @@
...
+   - The Code Map now provides a mode for the graph semantic model which disables collection of init delegates.
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:17]: This should be changed to something like this:
The Code Map performance was greatly optimized. The time for the construction of Code Map decreased from ~3-5 minutes to 1-2 seconds on code files with ~12 000 lines of code

File: docs/ReleaseNotes.md [hunk 6]
```diff
@@ -1,6 +1,28 @@
...
+
+
+### Fixed Bugs
+In this version of Acuminator, the following bugs have been fixed:
+ - The [PX1015](diagnostics/PX1015.md) diagnostic did not support counting of fluent BQL-like query parameters.
+ - The Code Map window frequently became out of sync with active document due to various reasons.
+ - The diagnostic suppression code fix was incorrectly displayed for multiple diagnostics on the same node.
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:24]: I would rephrase it:
The error suppression code action was incorrectly displayed when multiple diagnostics reported error on the same syntax node.

File: docs/diagnostics/PX1016.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# PX1016
+This document describes the PX1016 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                                                                                                      | Type                           | Code Fix    | 
+| ------ | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ----------- | 
+| PX1016 | A DAC or graph extension must include the public static IsActive method with the bool return type. Extensions which are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Available |
```
Comment (SENya1990) [inline docs/diagnostics/PX1016.md:8]: I would add that IsActive method must be paremeterless

File: docs/diagnostics/PX1033.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+This document describes the PX1033 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1033 | The DAC does not have an explicit primary key declaration | Warning (ISV Level 3: Informational) | Unavailable | 
+
+## Diagnostic Description
+You can define primary key in DACs. The diagnostic suggests adding primary key definition to a DAC. The warning is shown only for DACs that have either PXCacheName or PXPrimaryGraph attribute.
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:11]: There is a couple of other restrictions. Diagnostic will not check abstract DACs and DACs which has only unbound DAC properties (so-called "virtual" DACs, used for custom popups and filters for inquiry screens)

Comment (SENya1990) [reply docs/diagnostics/PX1033.md:11]: We usually add code fix description to the documentation. Here it can be useful to emphasize that code fix will generate not only primary key but also a Find method which will make PK more useful (and it is recommended to declare it)

File: docs/diagnostics/PX1034.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+This document describes the PX1034 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1034 | The DAC does not have an explicit foreign key declaration | Warning (ISV Level 3: Informational) | Unavailable | 
+
+## Diagnostic Description
+You can define foreign key in DACs. The diagnostic suggests adding foreign key definition to a DAC. The warning is shown only for DACs that have either PXCacheName or PXPrimaryGraph attribute.
```
Comment (SENya1990) [inline docs/diagnostics/PX1034.md:11]: There is a couple of other restrictions. Diagnostic will not check abstract DACs and DACs which has only unbound DAC properties (so-called "virtual" DACs, used for custom popups and filters for inquiry screens)

File: docs/diagnostics/PX1033.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# PX1033
+This document describes the PX1033 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1033 | The DAC does not have an explicit primary key declaration | Warning (ISV Level 3: Informational) | Unavailable |
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:8]: Code Fix is available for this diagnostic

File: docs/diagnostics/PX1034.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
+# PX1034
+This document describes the PX1034 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1034 | The DAC does not have an explicit foreign key declaration | Warning (ISV Level 3: Informational) | Unavailable |
```
Comment (SENya1990) [inline docs/diagnostics/PX1034.md:8]: Code Fix is available for this diagnostic

File: docs/diagnostics/PX1034.md (hunk 9 above)
Comment (SENya1990) [reply docs/diagnostics/PX1034.md:11]: Code fix should be described. For this diagnostic code fix will add public static class FK.
Inside it will place a big comment with examples of different foreign keys and a list of DAC properties which have a high probability of holding foreign key. This is determined by the analysis of attributes declared on DAC property. We look for the following attribute or attributes derived from them:
- PXParent 
- PXDBDefault 
- PXSelector 
- PXDimensionSelector 
- PXForeignReference

File: docs/diagnostics/PX1035.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# PX1035
+This document describes the PX1035 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1035 | The DAC has multiple primary key declarations | Warning (ISV Level 3: Informational) | Unavailable |
```
Comment (SENya1990) [inline docs/diagnostics/PX1035.md:8]: Code Fix is available for this diagnostic

File: docs/diagnostics/PX1035.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+This document describes the PX1035 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1035 | The DAC has multiple primary key declarations | Warning (ISV Level 3: Informational) | Unavailable | 
+
+## Diagnostic Description
+A DAC cannot have multiple primary key declarations. The warning is shown only for DACs that have either PXCacheName or PXPrimaryGraph attribute.
```
Comment (SENya1990) [inline docs/diagnostics/PX1035.md:11]: The same comment about abstract and fully unbound DACs as for PX1033/PX1034

File: docs/diagnostics/PX1035.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+To fix the issue, you should remove all but one primery key declarations.
```
Comment (SENya1990) [inline docs/diagnostics/PX1035.md:13]: The code fix should be described. It removes all primary key declarations except the selected one

File: docs/diagnostics/PX1036.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+# PX1036
+This document describes the PX1036 diagnostic.
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1036 | The DAC primary key should be named "PK". The class containing DAC foreign keys should be named "FK".  | Warning (ISV Level 3: Informational) | Unavailable |
```
Comment (SENya1990) [inline docs/diagnostics/PX1036.md:8]: The code fix is available

File: docs/diagnostics/PX1036.md [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1036 | The DAC primary key should be named "PK". The class containing DAC foreign keys should be named "FK".  | Warning (ISV Level 3: Informational) | Unavailable | 
+
+## Diagnostic Description
+The diagnostic checks the correct naming of key field declarations in DACs. For a primary key, the name should be PK. For foreign keys, the name of the static class which contains the foreign key classes must be FK.
+The warning is shown only for DACs that have either PXCacheName or PXPrimaryGraph attribute.
```
Comment (SENya1990) [inline docs/diagnostics/PX1036.md:12]: The same comment about abstract and fully unbound DACs as for PX1033/PX1034

File: docs/diagnostics/PX1036.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+To fix the issue, change the name of the primary key class to PK, the name of the class that contains foreign key classes to FK.
```
Comment (SENya1990) [inline docs/diagnostics/PX1036.md:14]: Currently only the check and code fix for PK is implemented although I'm not sure if it should be mentioned.
The code fix for PK changes the name of the primary key to PK. Imho it should be described

File: docs/ReleaseNotes.md [hunk 18]
```diff
@@ -6,22 +6,27 @@ Acuminator 2.3 includes the bug fixes and enhancements described in this section
...
-   - The following complementary information is now displayed in the Code Map about a graph view member: Whether the view is for mass processing operations, for setup form, or for filtering, whether the view is derived fom the base PXSelect or PXSelectReadOnly types.
-   - The Code Map now provides a mode for the graph semantic model which disables collection of init delegates.
+   - The Code Map now indicates whether each graph is for mass processing operations and whether each graph node corresponds to a graph or a graph extension.
+   - The Code Map now indicates whether each DAC field is auto-incremented (has the `AutoNumberAttribute` attribute).
+   - The Code Map performance was greatly optimized. The time for the construction of the Code Map decreased from 3-5 minutes to 1-2 seconds on code files with ~12 000 lines of code
+ - The Code Map now displays complementary information about graph views. A developer can see indicators for the following:
+   - Views for mass processing operations
+   - Setup views
+   - Filter views
+   - Custom views derived fron the base `PXSelect` types
```
Comment (SENya1990) [inline docs/ReleaseNotes.md:21]: a mistype - fro**m**

File: docs/diagnostics/PX1016.md [hunk 19]
```diff
@@ -5,12 +5,12 @@ This document describes the PX1016 diagnostic.
 
 | Code   | Short Description                                                                                                                      | Type                           | Code Fix    | 
 | ------ | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ----------- | 
-| PX1016 | A DAC or graph extension must include the public static IsActive method with the bool return type. Extensions which are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Available | 
+| PX1016 | A DAC or graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Available |
```
Comment (SENya1990) [inline docs/diagnostics/PX1016.md:8]: Code fix is unavailable for this diagnostic

File: docs/diagnostics/PX1033.md [hunk 20]
```diff
@@ -5,12 +5,16 @@ This document describes the PX1033 diagnostic.
...
 | Code   | Short Description                                 | Type  | Code Fix    | 
 | ------ | ------------------------------------------------- | ----- | ----------- | 
-| PX1033 | The DAC does not have an explicit primary key declaration | Warning (ISV Level 3: Informational) | Unavailable | 
+| PX1033 | The DAC does not have an explicit primary key declaration | Warning (ISV Level 3: Informational) | Available | 
 
 ## Diagnostic Description
-You can define primary key in DACs. The diagnostic suggests adding primary key definition to a DAC. The warning is shown only for DACs that have either PXCacheName or PXPrimaryGraph attribute.
+You can define a primary key field in a DAC. The diagnostic suggests adding a primary key definition to a DAC. 
+The warning is shown only for DACs that have the `PXCacheName` or `PXPrimaryGraph` attribute.
+The diagnostic does not check abstract DACs and DACs which have only unbound DAC properties (so-called "virtual" DACs used for custom popups and filters for inquiry forms).
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:13]: I'm not sure about including "virtual" DAC term into official documentation. It is not a well established term, which is used sometimes by our internal developers. There is also term **fully-unbound** which I think better from transfers meaning. 
Or we can avoid inclusion of any term at all

File: docs/diagnostics/PX1033.md [hunk 21]
```diff
@@ -5,12 +5,16 @@ This document describes the PX1033 diagnostic.
...
 
-## Example of Code For Which the Warning Is Shown
+The code fix generates a primary key and the 'Find' method which makes PK more useful.
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:15]: I would rephrase a bit - "which allows to easily find DAC row by providing values composing primary key ".

File: docs/diagnostics/PX1016.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+{
+    ...
+	// no isActive method
+}
+```
+
+## Example of the Fixed Code
+
+```C#
+// A diagnostic must be shown on SOOrderEntryExt
```
Comment (SENya1990) [inline docs/diagnostics/PX1016.md:29]: the comment is incorrect

File: docs/diagnostics/PX1016.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+## Example of the Fixed Code
+
+```C#
+// A diagnostic must be shown on SOOrderEntryExt 
+public class SOOrderEntryExt : PXGraphExtension<SOOrderEntry>
+{
+    ...
+	public static Boolean IsActive() => true;
```
Comment (dropsonic) [inline docs/diagnostics/PX1016.md:33]: The example is not correct. Always returning `true` doesn't fix the issue, it just suppresses the diagnostic.

**Comment by EAndrosova** [comment]
Updated RN and diagnostic descriptions according to comments, please review

File: docs/diagnostics/PX1016.md [hunk 24]
```diff
@@ -5,7 +5,7 @@ This document describes the PX1016 diagnostic.
 
 | Code   | Short Description                                                                                                                      | Type                           | Code Fix    | 
 | ------ | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ----------- | 
-| PX1016 | A DAC or graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Available | 
+| PX1016 | A DAC or graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Unavailable | 
 
 ## Diagnostic Description
```
Comment (dropsonic) [inline docs/diagnostics/PX1016.md:10]: Please also add the information that if the developer intentionally wants to make an extension "always-on", they must suppress the diagnostic by using a suppression comment. We got a question about it on Stack Overflow.

Comment (SENya1990) [reply docs/diagnostics/PX1016.md:10]: Link to the question
https://stackoverflow.com/questions/63050531/acumatica-acuminator-error-px1016-looking-for-more-information-regarding-this/63057097#63057097

File: docs/diagnostics/PX1033.md [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+```C#
+[PXCacheName(Messages.SOOrder)]
+public class SOOrder : IBqlTable
+{
+	#region OrderType
+	public abstract class orderType : PX.Data.BQL.BqlString.Field<orderType> { }
+	protected String _OrderType;
+	[PXDBString(2, IsKey = true, IsFixed = true, InputMask = ">aa")]
+	...
+	public virtual String OrderType
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:30]: The example is a bit trivial, we can use more close to real life example by adding OrderNbr - this way you will have a real SOOrder key. 
Also, it is very outdated style - currently everyone uses auto-properties in C# instead of explicit backing field, so the example could be rewritten like this:
```csharp
[PXCacheName(Messages.SOOrder)]
public class SOOrder : IBqlTable
{
	#region OrderType
	public abstract class orderType : PX.Data.BQL.BqlString.Field<orderType> { }

	[PXDBString(2, IsKey = true, IsFixed = true, InputMask = ">aa")]
	...
	public virtual String OrderType {get; set; }
	#endregion
        
        #region OrderNbr
	[PXDBString(IsKey = true, InputMask = "")]
	[PXDefault]
	[PXUIField(DisplayName = "Order Nbr.")]
	public string OrderNbr { get; set; }

	public abstract class orderNbr : IBqlField { }
	#endregion
}
```

File: docs/diagnostics/PX1035.md [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1035 | The DAC has multiple primary key declarations | Warning (ISV Level 3: Informational) | Available | 
+
+## Diagnostic Description
+A DAC cannot have multiple primary key declarations. The warning is shown only for DACs that have the `PXCacheName` or `PXPrimaryGraph` attribute.
+The diagnostic does not check abstract DACs and DACs which have only unbound DAC properties (so-called "virtual" DACs used for custom popups and filters for inquiry forms).
+
+To fix the issue, you should remove all primery key declarations except one.
```
Comment (SENya1990) [inline docs/diagnostics/PX1035.md:14]:
```suggestion
To fix the issue, you should remove all primary key declarations except one.
```

File: docs/diagnostics/PX1035.md [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+
+## Summary
+
+| Code   | Short Description                                 | Type  | Code Fix    | 
+| ------ | ------------------------------------------------- | ----- | ----------- | 
+| PX1035 | The DAC has multiple primary key declarations | Warning (ISV Level 3: Informational) | Available | 
+
+## Diagnostic Description
+A DAC cannot have multiple primary key declarations. The warning is shown only for DACs that have the `PXCacheName` or `PXPrimaryGraph` attribute.
+The diagnostic does not check abstract DACs and DACs which have only unbound DAC properties (so-called "virtual" DACs used for custom popups and filters for inquiry forms).
```
Comment (SENya1990) [inline docs/diagnostics/PX1035.md:12]: @dropsonic, do we use term "virtual" for DACs in official documentation? Is it correct to use it or the term "fully-unbound" DAC?

Comment (dropsonic) [reply docs/diagnostics/PX1035.md:12]: I think "unbound" is more correct; the best way is to add a proper clarification ("DAC that is not bound to any DB table and exists only in-memory").

**Comment by EAndrosova** [comment]
Updated and removed "virtual" from all diagnostic descriptions

File: docs/diagnostics/PX1033.md [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+	public virtual String OrderType {get; set; }
+	#endregion
+        
+    #region OrderNbr
+	[PXDBString(IsKey = true, InputMask = "")]
+	[PXDefault]
+	[PXUIField(DisplayName = "Order Nbr.")]
+	public string OrderNbr { get; set; }
+
+	public abstract class orderNbr : IBqlField { }
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:69]: to be consistent in examples, orderNbr should inherit from BqlString:
```suggestion
	public abstract class orderNbr : : PX.Data.BQL.BqlString.Field<orderNbr > { }
```

File: docs/diagnostics/PX1033.md [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+	public virtual String OrderType {get; set; }
+	#endregion
+        
+        #region OrderNbr
+	[PXDBString(IsKey = true, InputMask = "")]
+	[PXDefault]
+	[PXUIField(DisplayName = "Order Nbr.")]
+	public string OrderNbr { get; set; }
+
+	public abstract class orderNbr : IBqlField { }
```
Comment (SENya1990) [inline docs/diagnostics/PX1033.md:39]: to be consistent in examples, orderNbr should inherit from BqlString:
```suggestion
	public abstract class orderNbr : : PX.Data.BQL.BqlString.Field<orderNbr > { }
```

File: docs/diagnostics/PX1034.md (hunk 11 above)
Comment (SENya1990) [reply docs/diagnostics/PX1034.md:8]: @EAndrosova  please fix the value for Code Fix column in the table

Comment (SENya1990) [reply docs/diagnostics/PX1034.md:8]:
```suggestion
| PX1034 | The DAC does not have an explicit foreign key declaration | Warning (ISV Level 3: Informational) | Available | 
```

File: docs/diagnostics/PX1016.md (hunk 24 above)
Comment (EAndrosova) [reply docs/diagnostics/PX1016.md:10]: @dropsonic  I have added the following paragraph to the diagnostic description
In case you do not want to define the `IsActive` method as suggested above, you should suppress the diagnostic for better performance, otherwise the `IsActive` method will be called constantly to check whether the extension is enabled.

File: docs/diagnostics/PX1016.md [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+## Summary
+
+| Code   | Short Description                                                                                                                      | Type                           | Code Fix    | 
+| ------ | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ----------- | 
+| PX1016 | A DAC or graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Unavailable | 
+
+## Diagnostic Description
+A DAC or a graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance.
+
+To fix the issue, in the `IsActive` method, you should enable the DAC or graph extension if the feature for which this extension is needed is enabled.
```
Comment (dropsonic) [inline docs/diagnostics/PX1016.md:13]: The extension can be enabled or disabled by any condition, not necessary by a feature.

File: docs/diagnostics/PX1016.md [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+| ------ | -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ----------- | 
+| PX1016 | A DAC or graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance. Suppress the error if you need the DAC extension to be constantly active. | Error | Unavailable | 
+
+## Diagnostic Description
+A DAC or a graph extension must include the public static `IsActive` method with no parameters and the `bool` return type. Extensions that are constantly active reduce performance.
+
+To fix the issue, in the `IsActive` method, you should enable the DAC or graph extension if the feature for which this extension is needed is enabled. 
+
+In case you do not want to define the `IsActive` method as suggested above, you should suppress the diagnostic for better performance, otherwise the `IsActive` method will be called constantly to check whether the extension is enabled.
```
Comment (dropsonic) [inline docs/diagnostics/PX1016.md:15]: _"In case you want your extension to be always enabled ..."_
Also, suppressing the diagnostic doesn't affect performance. Declaring `IsActive()` method does.

File: docs/diagnostics/PX1016.md [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+```C#
+// A diagnostic must be shown on SOOrderEntryExt 
+public class SOOrderEntryExt : PXGraphExtension<SOOrderEntry>
+{
+    ...
+	// no isActive method
+}
+```
+
+## Example of the Fixed Code
```
Comment (dropsonic) [inline docs/diagnostics/PX1016.md:28]: Please also add an example with a suppression comment:

```
// Acuminator disable once PX1016 ExtensionDoesNotDeclareIsActiveMethod because this extension should be always enabled
public class SOOrderEntryExt : PXGraphExtension<SOOrderEntry>
{
    ...
}
```

File: docs/diagnostics/PX1033.md [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+The diagnostic does not check abstract DACs and DACs which have only unbound DAC properties (fully-unbound DACs used for custom popups and filters for inquiry forms).
+
+To fix the issue, add the PK class to the DAC.
+
+The code fix generates a primary key and the 'Find' method which allows to easily find DAC row by providing values composing primary key.
+
+## Example of Code for Which the Warning Is Shown
+
+```C#
+[[PXCacheName(Messages.SOOrder)]
```
Comment (dropsonic) [inline docs/diagnostics/PX1033.md:22]: Extra `[` symbol at the beginning

**Comment by EAndrosova** [comment]
@dropsonic updated px1016 and px1033 according to your comments, please review

**Comment by dropsonic** [comment]
@EAndrosova are we waiting for the editor review or we can merge the PR?

**Comment by EAndrosova** [comment]
@dropsonic you can merge, I already processed editor's comments

**Comment by dropsonic** [comment]
@EAndrosova thanks!
